components/layout_2020/flow/construct.rs

```diff
@@ -214,34 +214,120 @@ impl BlockContainer {
+    ) -> Self {
+        let text_decoration_line =
+            propagated_text_decoration_line | info.style.clone_text_decoration_line();
+        BlockContainerBuilder {
+            context,
+            info,
+            block_level_boxes: Vec::new(),
+            ongoing_inline_formatting_context: InlineFormattingContext::new(
+                text_decoration_line,
+                /* has_first_formatted_line = */ true,
+                /* ends_with_whitespace */ false,
+            ),
+            ongoing_inline_boxes_stack: Vec::new(),
+            anonymous_style: None,
+            anonymous_table_content: Vec::new(),
+        }
+    }
+
+    pub(crate) fn finish(mut self) -> BlockContainer {
+        debug_assert!(self.ongoing_inline_boxes_stack.is_empty());
 
-        debug_assert!(builder.ongoing_inline_boxes_stack.is_empty());
+        self.finish_anonymous_table_if_needed();
 
-        if !builder.ongoing_inline_formatting_context.is_empty() {
-            if builder.block_level_boxes.is_empty() {
+        if !self.ongoing_inline_formatting_context.is_empty() {
+            if self.block_level_boxes.is_empty() {
                 return BlockContainer::InlineFormattingContext(
-                    builder.ongoing_inline_formatting_context,
+                    self.ongoing_inline_formatting_context,
                 );
             }
-            builder.end_ongoing_inline_formatting_context();
+            self.end_ongoing_inline_formatting_context();
         }
 
-        let block_level_boxes = if context.use_rayon {
-            builder
-                .block_level_boxes
+        let context = self.context;
+        let block_level_boxes = if self.context.use_rayon {
+            self.block_level_boxes
                 .into_par_iter()
                 .map(|block_level_job| block_level_job.finish(context))
                 .collect()
         } else {
-            builder
-                .block_level_boxes
+            self.block_level_boxes
                 .into_iter()
                 .map(|block_level_job| block_level_job.finish(context))
                 .collect()
         };
 
         BlockContainer::BlockLevelBoxes(block_level_boxes)
     }
+
+    fn finish_anonymous_table_if_needed(&mut self) {
+        if self.anonymous_table_content.is_empty() {
+            return;
+        }
+
+        // Text decorations are not propagated to atomic inline-level descendants.
+        let inline_table = !self.ongoing_inline_boxes_stack.is_empty();
+        let propagated_text_decoration_line = if inline_table {
+            TextDecorationLine::NONE
+        } else {
+            self.ongoing_inline_formatting_context.text_decoration_line
+        };
+
+        let contents: Vec<AnonymousTableContent<'dom, Node>> =
+            self.anonymous_table_content.drain(..).collect();
+        let last_text = match contents.last() {
+            Some(AnonymousTableContent::Text(info, text)) => Some((info.clone(), text.clone())),
```

Comment:
Instead of cloning, maybe just `pop()` the last item from `contents`?

Reply:
`pop()` here would be a little awkward, because if it's not text, we'd either have to push it back on or process it specially below (after matching again). I think this might be the cleanest way to do it after investigating. Cloning the text in this case shouldn't be too expensive, because it's behind a `Cow` smart-pointer.

Reply:
It also seems weird to handle the same node in 2 places, even if it ends up being fine because `construct_anonymous` ignores texts.

So even with the current approach, it could be good to pop.